crypto/rand/rand_uniform.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "crypto/rand.h"
+#include "internal/common.h"
+
+/*
+ * Implementation an optimal random integer in a range function.
+ *
+ * Essentially it boils down to incrementally generating a fixed point
+ * number on the interval [0, 1) and multiplying this number by the upper
+ * range limit.  Once it is certain what the fractional part contributes to
+ * the integral part of the product, the algorithm has produced a definitive
+ * result.
+ *
+ * Refer: https://github.com/apple/swift/pull/39143 for a fuller description
+ * of the algorithm.
+ */
+uint32_t ossl_rand_uniform_uint32(OSSL_LIB_CTX *ctx, uint32_t upper, int *err)
+{
+    uint32_t i, f;      /* integer and fractional parts */
+    uint32_t f2, rand;  /* extra fractional part and random material */
+    uint64_t prod;      /* temporary holding double width product */
+    const int max_followup_iterations = 10;
+    int j;
+
+    if (!ossl_assert(upper > 0)) {
+        *err = 0;
+        return 0;
+    }
+    if (unlikely(upper == 1))
+        return 0;
+
+    /* Get 32 bits of entropy */
+    if (RAND_bytes_ex(ctx, (unsigned char *)&rand, sizeof(rand), 0) <= 0) {
+        *err = 1;
+        return 0;
+    }
+
+    /*
+     * We are generating a fixed point number on the interval [0, 1).
+     * Multiplying this but the range gives us a number on [0, upper).
+     * The high word of the multiplication result represents the integral
+     * part we want.  The lower word is the fractional part.  We can early exit if
+     * the fractional part is small enough that no carry from the next lower
+     * word can cause an overflow and carry into the integer part.  This
+     * happens when the fractional part is bounded by 2^32 - upper which
+     * can be simplified to just -upper (as an unsigned integer).
+     */
+    prod = (uint64_t)upper * rand;
+    i = prod >> 32;
+    f = prod & 0xffffffff;
+    if (likely(f <= 1 + ~upper))    /* 1+~upper == -upper but compilers whine */
+        return i;
+
+    /*
+     * We're in the position where the carry from the next word *might* cause
+     * a carry to the integral part.  The process here is to generate the next
+     * word, multiply it by the range and add that to the current word.  If
+     * it overflows, the carry propagates to the integer part (return i+1).
+     * If it can no longer overflow regardless of further lower order bits,
+     * we are done (return i).  If there is still a chance of overflow, we
+     * repreat the process with the next lower word.
```

Comment:
repreat -> repeat

Reply:
Hmmmm....I ran this through a spell checker....

All fixed now.